wfl/calculators/castep.py

```diff
@@ -1,133 +1,134 @@
+        "keep_files") if needed.  For example, directory on a local disk with fast file I/O.
+    calculator_exec: str
+        command for Castep, without any prefix or redirection set.
+        for example: "mpirun -n 4 castep.mpi"
+        Alternative for "castep_command", for consistency with other wfl calculators. 
+
+    **kwargs: arguments for ase.calculators.Castep.Castep
     """
-    # use list of atoms in any case
-    at_list = atoms_to_list(atoms)
-
-    # default properties
-    if properties is None:
-        properties = __default_properties
-
-    # keyword setup
-    if calculator_kwargs is None:
-        calculator_kwargs = dict()
-
-    if workdir_root is None:
-        # using the current directory
-        workdir_root = os.getcwd()
-    else:
-        pathlib.Path(workdir_root).mkdir(parents=True, exist_ok=True)
-
-    # set up the keywords of the calculator -- this is reused
-    kwargs_this_calc = dict(
-        calculator_kwargs, label=calculator_kwargs.get("label", "castep"),
-    )
-    if calculator_command is not None:
-        kwargs_this_calc["castep_command"] = calculator_command
-
-    if kwargs_this_calc.get("castep_pp_path", None) is not None:
-        # make sure we are looking for pspot if path given
-        kwargs_this_calc["find_pspots"] = True
-
-    for at in at_list:
-        # periodic calculations
-        nonperiodic, properties_use = handle_nonperiodic(at, properties)
-        kwargs_this_calc["calculate_stress"] = "stress" in properties_use
-
-        # create temp dir and calculator
-        rundir = tempfile.mkdtemp(dir=workdir_root, prefix=dir_prefix)
-        at.calc = Castep(directory=rundir, **kwargs_this_calc)
-
-        # calculate
-        calculation_succeeded = False
+    implemented_properties = ["energy", "forces", "stress"]
+
+    # new default value of num_inputs_per_python_subprocess for calculators.generic,
+    # to override that function's built-in default of 10
+    wfl_generic_def_autopara_info = {"num_inputs_per_python_subprocess": 1}
+
+    def __init__(self, keep_files="default", rundir_prefix="run_CASTEP_",
+                 workdir=None, scratchdir=None,
+                 calculator_exec=None, **kwargs):
+
+        kwargs = deepcopy(kwargs)
+        if calculator_exec is not None:
+            if "castep_command" in kwargs:
+                raise ValueError("Cannot specify both calculator_exec and command")
+            kwargs["castep_command"] = calculator_exec 
+
+        if kwargs.get("castep_pp_path", None) is None:
+            # make sure we are looking for pspot if path given
+            kwargs["find_pspots"] = True
+
+        # WFLFileIOCalculator is a mixin, will call remaining superclass constructors for us
+        super().__init__(keep_files=keep_files, rundir_prefix=rundir_prefix,
+                         workdir=workdir, scratchdir=scratchdir, **kwargs)
+
+        self.directory = self._directory
```

Comment:
I think it's worth commenting where stock Castep is non-standard.  Can we add something like 
```
# Copy from Castep's `_directory` attr to more standard `directory` attr so the rest of `wfl` can find it
```
I'll try to find the other places where I think such comments are worth it.